### PR TITLE
spinlock: spin_try_lock refinement

### PR DESCRIPTION
**Pull request by bkokoszx**
Use block expression instead of do {} while(0) macro
in spin_try_lock in order to return its value in a
safer way.

Signed-off-by: Bartosz Kokoszko <bartoszx.kokoszko@linux.intel.com>

### CODE REVIEW

File: src/include/sof/spinlock.h [hunk 1]
```diff
@@ -82,14 +82,14 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
```
Comment (dbaluta) [inline src/include/sof/spinlock.h]: @bkokoszx care to explain why this change is better then current implemention?

File: src/include/sof/spinlock.h [hunk 2]
```diff
@@ -82,14 +82,14 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
 
 /* all SMP spinlocks need init, nothing todo on UP */
 #define spinlock_init(lock) \
-	do { \
+	({ \
 		arch_spinlock_init(lock); \
 		(*lock)->user = __LINE__; \
-	} while (0)
+	})
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Do we need to return `(*lock)->user`? I doubt it.

File: src/include/sof/spinlock.h [hunk 3]
```diff
@@ -101,11 +101,11 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
 			trace_lock_value((lock)->user); \
 			panic(SOF_IPC_PANIC_DEADLOCK); /* lock not acquired */ \
 		} \
-	} while (0)
+	})
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: GNU block expressions aren't useful when the last thing is a for loop; that is either void or not something that can be made an expression. And since this one only returns on success, there isn't need for this one either.

File: src/include/sof/spinlock.h [hunk 4]
```diff
@@ -118,19 +118,19 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
 					lock_dbg_user[__i]); \
 			} \
 		} \
-	} while (0)
+	})
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Ditto, not something that is to be used as an expression.

File: src/include/sof/spinlock.h [hunk 5]
```diff
@@ -118,19 +118,19 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
...
 
 #define spin_lock_dbg() \
-	do { \
+	({ \
 		trace_lock("LcE"); \
 		trace_lock_value(__LINE__); \
-	} while (0)
+	})
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Nor is this.

File: src/include/sof/spinlock.h [hunk 6]
```diff
@@ -118,19 +118,19 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
...
 
 #define spin_unlock_dbg() \
-	do { \
+	({ \
 		trace_lock("LcX"); \
 		trace_lock_value(__LINE__); \
-	} while (0)
+	})
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Nor this.

File: src/include/sof/spinlock.h [hunk 7]
```diff
@@ -184,36 +184,36 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
...
 		arch_spin_lock(lock); \
-	} while (0)
+	})
 
 #define spin_try_lock(lock, ret) \
-	do { \
+	({ \
 		spin_lock_dbg(); \
 		ret = arch_try_lock(lock); \
-	} while (0)
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: This is the only function where this change would have benefitted, and you still used an out parameter instead of actually returning it.

**Comment by jajanusz** [comment]
Instead of gnu extension that returns value from block I'd prefer to have it refactored to be inline functions (to have f.e. type of lock that is often misused as pointer instead of pointer to pointer).

**Comment by paulstelian97** [comment]
> Instead of gnu extension that returns value from block I'd prefer to have it refactored to be inline functions (to have f.e. type of lock that is often misused as pointer instead of pointer to pointer).

I have already posted the issue in fact. Sure, that would be the best but until that block expressions I argue are acceptable.

**Comment by jajanusz** [comment]
> > Instead of gnu extension that returns value from block I'd prefer to have it refactored to be inline functions (to have f.e. type of lock that is often misused as pointer instead of pointer to pointer).
> 
> I have already posted the issue in fact. Sure, that would be the best but until that block expressions I argue are acceptable.

What do block expression give us? It's not like switch from macros to expressions is free, you already change macros to something that returns value, so it can also be function instead of nontyped expression.

**Comment by paulstelian97** [comment]
> > > Instead of gnu extension that returns value from block I'd prefer to have it refactored to be inline functions (to have f.e. type of lock that is often misused as pointer instead of pointer to pointer).
> > 
> > 
> > I have already posted the issue in fact. Sure, that would be the best but until that block expressions I argue are acceptable.
> 
> What do block expression give us? It's not like switch from macros to expressions is free, you already change macros to something that returns value, so it can also be function instead of nontyped expression.

Fair enough. @bkokoszx Can you resubmit with this change? You'd use `static inline` for each of these functions.

**Comment by bkokoszx** [comment]
@jajanusz @paulstelian97 

Thanks for review. I've made this change a bit thoughtlessly. I will resubmit with inline functions.

File: src/include/sof/spinlock.h [hunk 8]
```diff
@@ -81,44 +81,46 @@ extern uint32_t lock_dbg_atomic;
...
-		if (__tries == 0) { \
-			trace_lock_error("DED"); \
-			trace_lock_value(__LINE__); \
-			trace_lock_value((lock)->user); \
-			panic(SOF_IPC_PANIC_DEADLOCK); /* lock not acquired */ \
-		} \
-	} while (0)
+static inline void spin_try_lock_dbg(spinlock_t *lock)
+{
+	int __tries;
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: No longer need to prefix variables with __; that was only necessary in macros.

**Comment by bkokoszx** [comment]
@jajanusz @paulstelian97 
I've changed spinlocks to ```static inline``` functions. I had to forward declare two interrupt inline functions in ```sof\spinlock.h```.
@lgirdwood 
Is there a specific reason why we use typedef:
```
typedef struct {
	volatile uint32_t lock;
#if CONFIG_DEBUG_LOCKS
	uint32_t user;
#endif
} spinlock_t;
```
in ```arch\spinlock.h``` instead of simple struct (It seems to me there is no possibility to "forward declare" typedefs)?

File: src/include/sof/spinlock.h [hunk 9]
```diff
@@ -81,44 +81,46 @@ extern uint32_t lock_dbg_atomic;
...
+
+	for (__tries = DBG_LOCK_TRIES; __tries > 0; __tries--) {
+		if (arch_try_lock(lock))
+			break;	/* lock acquired */
+	}
+	if (__tries == 0) {
+		trace_lock_error("DED");
+		trace_lock_value(__LINE__);
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Oof, line numbers will be messed up when using inline functions. More thought needs to go into this. For now you can just remove the line number because it'll be incorrect anyway.

File: src/include/sof/spinlock.h [hunk 10]
```diff
@@ -81,44 +81,46 @@ extern uint32_t lock_dbg_atomic;
...
-	} while (0)
+static inline void spin_lock_log(spinlock_t *lock)
+{
+	if (lock_dbg_atomic) {
+		int __i = 0;
+		int  __count = lock_dbg_atomic >= DBG_LOCK_USERS ?
+			DBG_LOCK_USERS : lock_dbg_atomic;
+
+		trace_lock_error("eal");
+		trace_lock_value(__LINE__);
```
Comment (paulstelian97) [inline src/include/sof/spinlock.h]: Same complaint about line numbers.

**Comment by paulstelian97** [comment]
My two cents on the line number issue: you can wrap the new spin_lock_* functions inside macros which pass as a "hidden" second parameter the line number for debug. Such simpler macros used just for this purpose are definitely more acceptable to me than what we have already. @jajanusz What do you think about that?

Also you could use the CI to see that there are some spinlock usages that haven't been updated in i.MX and HOST.

**Comment by jajanusz** [comment]
> My two cents on the line number issue: you can wrap the new spin_lock_* functions inside macros which pass as a "hidden" second parameter the line number for debug. Such simpler macros used just for this purpose are definitely more acceptable to me than what we have already. @jajanusz What do you think about that?
> 
> Also you could use the CI to see that there are some spinlock usages that haven't been updated in i.MX and HOST.

Yea, what you wrote is good. When there is __LINE__ needed then macro should be proxy that adds extra value to inline function.

File: src/drivers/dw/dma.c [hunk 11]
```diff
@@ -135,7 +135,7 @@ static struct dma_chan_data *dw_dma_channel_get(struct dma *dma,
 	trace_dwdma("dw_dma_channel_get(): dma %d request channel %d",
 		    dma->plat_data.id, req_chan);
 
-	spin_lock_irq(dma->lock, flags);
+	flags = spin_lock_irq(dma->lock);
```
Comment (dbaluta) [inline src/drivers/dw/dma.c]: My eyes hurt looking at this. I do understand the need for replacing macros with functions, but I don't really like that we change the prototype.

Would it be possible to pass a pointer to flags as arguments, so that at least people used with locking API in kernel won't be confused about this.

So, instead of 
```flags = spin_lock_irq(lock)```
you will have
```
spin_lock_irq(lock, &flags);
```

Comment (jajanusz) [reply src/drivers/dw/dma.c]: @dbaluta I don't like passing something that looks like passing by value, but in reality you pass it like reference and it gets modified (that's how old macros worked). How it in in kernel usually? It is like our old macros or with pointer? (pointer is ok for me, at least I can see that flags gonna be modified)

Comment (dbaluta) [reply src/drivers/dw/dma.c]: @jajanusz in kernel is like our old approach with macros. What problems are we having with macros approach so that we need to change it, now?

**Comment by dbaluta** [comment]
@bkokoszx this PR is a big NACK from me. 

Just look how peope are supposed to use the API:

```
flags = spin_lock_irq(channel->dma->lock);
spin_unlock_irq(channel->dma->lock, flags);
```

Also, you don't provide a reason on why do we need to change this. Please don't bring-up the religious discussion that inline functions are better than macros.

File: src/drivers/dw/dma.c (hunk 11 above)
Comment (jajanusz) [reply src/drivers/dw/dma.c]: Problem - people sometimes pass wrong type of pointer and get no compilation warning and as I wrote previously you get nonexpected behaviour for your variables - you need to go one step deeper to understand code. That's what's wrong in this one case, I won't start religious war about why macros are even worse in other aspects  ;)

Comment (jajanusz) [reply src/drivers/dw/dma.c]: @dbaluta And if it's really what's in kernel and if they keep using it, I can live with that, but if they are moving away from that and it's only legacy stuff then I'd prefer to handle it differently

Comment (dbaluta) [reply src/drivers/dw/dma.c]: @jajanusz here is the implementation from Linux kernel:

https://elixir.bootlin.com/linux/latest/source/include/linux/spinlock.h#L247

Bottom line is that even they use macros they do some sort of type checking using this

https://elixir.bootlin.com/linux/latest/source/include/linux/typecheck.h#L9

Maybe we can do the same?

Comment (jajanusz) [reply src/drivers/dw/dma.c]: @dbaluta Looks like an improvment while keeping our codebase to be more familiar for kernel developers, so it's good.
@bkokoszx Please just add checks then

Comment (bkokoszx) [reply src/drivers/dw/dma.c]: @dbaluta 
Thanks for examples.
@lgirdwood 
Is is ok for you to add typechecks macros?

Comment (lgirdwood) [reply src/drivers/dw/dma.c]: Lets not redesign, lets do it the Linux way with typecheck macros too. We can benefit from how this is done in Linux where it's probably been debated to death (and also validated with lots of architectures and compilers).

**Comment by bkokoszx** [comment]
@jajanusz @dbaluta @paulstelian97 
I was thinking about typechecks a bit and I am not sure do we really need it. Spin locks 
invokes other inline functions (e.g. arch_spinlock_init(), arch_try_lock()) and they can be treated by
us as type checks.

@paulstelian97 
In this PR I come back to the beginning and change spin_try_lock() macro from do {} while(0) to block expression.

**Comment by bkokoszx** [comment]
@paulstelian97 
Yes, just updated

**Comment by lgirdwood** [comment]
CI not related to PR

File: src/drivers/dw/dma.c (hunk 11 above)
Comment (dbaluta) [reply src/drivers/dw/dma.c]: @lgirdwood @paulstelian97 @bkokoszx so looking again at the spin_lock implementation in Linux its not that easy to follow but:
* type check for lock parameter is done because spin_lock_ macro eventually goes to an inline function call.
* type check for flags is done with the help of typecheck macro

So, @anyone feel free to jump with ideas of implementing this.

Comment (lgirdwood) [reply src/drivers/dw/dma.c]: @dbaluta what's our current delta with the Linux method ? My understanding was the latest updated aligned to Linux ?

Comment (bkokoszx) [reply src/drivers/dw/dma.c]: @lgirdwood @dbaluta 
ATM we do not have type checks for ```flags```. Type check for ```lock``` is done, because in our example ```spin_lock_``` macro also goes to an inline function call.
